### PR TITLE
#91 web exception handler

### LINKED ISSUE #91: Implement Exception Handler
This Jira is to implement exception handler for the web part

Requirements:

1. Create new Annotation **ControllerErrorHandler** or different name it is up to you, it will mark the class that it will be responsible for error handling
2. Create new Annotation **ExceptionHandler** or different name it is up to you, it will mark the class method that it is responsible for specific exception value should be extend type of Throwable
4. Add new method **handleError** to **DispatcherServlet** that will be called if exception has been occurred. It should get as input httpreq, httpresp, throwable and call **processRequest(req, resp)** once error will be handled.

 
How it should work:

1. **BringContainer** creates new Bean for class that marked by **ControllerErrorHandler** annotation
2. On **DispatcherServlet** start up we retrieve all beans marked as **ControllerErrorHandler** from **BringContainer**
3. When exception occurs it propagates up the call stack, if it is not caught within the controller method it continues to propagate up to **DispatcherServlet**
4. In **DispatcherServlet** we will caught the exception and call **handleError** method
5. **handleError** will retrieve the last bean which has method public method that marked as **ExceptionHandler** and has value that match the exception
6. When required method has been found we should invoke it, and add output to response and then call **processRequest(req, resp)** again

### PR DESCRIPTION
**Pull request by s-barabash**

### CODE REVIEW

File: src/main/java/io/github/bobocodebreskul/context/annotations/ExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Indicates that the annotated method serves as an exception handler within a corresponding
+ * Controller. This annotation is intended for use in combination with the
+ * {@link ErrorHandlerController @ErrorHandlerController} annotation.
+ *
```
Comment (AlexandrKarpachov) [inline src/main/java/io/github/bobocodebreskul/context/annotations/ExceptionHandler.java]: twice writed ErrorHandlerController looks like a typo.

File: src/main/java/io/github/bobocodebreskul/server/DispatcherServlet.java [hunk 2]
```diff
@@ -75,82 +79,154 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
...
+      Method method = getMethod(controllerMethod);
+
+      Object result = method.invoke(controllerMethod.controller());
+      try (PrintWriter writer = resp.getWriter()) {
+        writer.println(mapper.writeValueAsString(result));
+        resp.setStatus(HttpServletResponse.SC_OK);
+        writer.flush();
+      }
+    } catch (Exception ex) {
+      if (ex instanceof InvocationTargetException) {
```
Comment (JoshukLite) [inline src/main/java/io/github/bobocodebreskul/server/DispatcherServlet.java:168]: I believe pattern matching could be used here

File: src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    if (Arrays.stream(methodParameterTypes)
+        .filter(Throwable.class::isAssignableFrom)
+        .count() != 1) {
+      throw new MethodValidationException("Only 1 exceptions is allowed for errorhandler method");
+    }
+
+    if (methodParameterTypes.length == 2 &&
+        Arrays.stream(methodParameterTypes)
+            .filter(methodType -> methodType.isInstance(HttpServletRequest.class))
+            .count() != 1) {
```
Comment (JoshukLite) [inline src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java]: What is expectation of this? Is it expected that user can inject `HttpServletRequest` into his error handling method?
If yes, then method `isInstatnce(HttpServletRequest.class` doesn't when HttpServletRequest object is injected to error handling method.

File: src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    return controllerMethodMap;
+  }
+
+  private static Class<?> getErrorClass(Class<?>[] methodParameterTypes) {
+    if (methodParameterTypes.length == 1) {
+      return methodParameterTypes[0];
+    }
+
+    if (methodParameterTypes.length == 2 && methodParameterTypes[0].isInstance(Throwable.class)) {
```
Comment (JoshukLite) [inline src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java]: It seems that any exception class will fail `.isInstance(Throwable.class)` check, so this method will return always 2nd argument if error handling method has 2 parameters.

File: src/main/java/io/github/bobocodebreskul/server/DispatcherServlet.java [hunk 5]
```diff
@@ -75,82 +79,154 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
...
+  }
+
+  private static Object getMethodInvokeResult(Method method, Object controller,
+      HttpServletRequest req, Throwable ex)
+      throws IllegalAccessException, InvocationTargetException {
+    if (Throwable.class.isAssignableFrom(method.getParameterTypes()[0])) {
+      return method.invoke(controller, ex, req);
     }
+    return method.invoke(controller, req, ex);
   }
```
Comment (AlexandrKarpachov) [inline src/main/java/io/github/bobocodebreskul/server/DispatcherServlet.java:231]: Need to take into account that user can set arguments in different order.

**Review by AlexandrKarpachov — CHANGES_REQUESTED** [review]

**Review by AlexandrKarpachov — APPROVED** [review]

**Review by mykolaFilimonov — APPROVED** [review]

File: src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    return container.getAllBeans()
+        .stream()
+        .filter(bean -> bean.getClass().isAnnotationPresent(ErrorHandlerController.class))
+        .toList();
+  }
+
+
+  private static void validateMethodParams(Class<?>[] methodParameterTypes) {
+    if (methodParameterTypes.length > 2 || methodParameterTypes.length == 0) {
+      throw new MethodValidationException("Invalid parameter quantaty");
```
Comment (golichenko02) [inline src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java]:
```suggestion
throw new MethodValidationException("Invalid parameter quantity");
```

File: src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    if (Arrays.stream(methodParameterTypes)
+        .filter(Throwable.class::isAssignableFrom)
+        .count() != 1) {
+      throw new MethodValidationException("Only 1 exceptions is allowed for errorhandler method");
```
Comment (golichenko02) [inline src/main/java/io/github/bobocodebreskul/server/WebErrorHandlerControllerScanner.java:102]:
```suggestion
throw new MethodValidationException("Only 1 exception is allowed for the error handler method");
```

**Review by golichenko02 — COMMENTED** [review]
Good job! Left minor comments

**Pull request merged**